src/libmongoc/tests/test-mongoc-server-stream.c

```diff
@@ -0,0 +1,227 @@
+   mock_server_replies_simple (request, HELLO_PRE_OPMSG);
+   request_destroy (request);
+   /* Check that the mock server receives an OP_QUERY. */
+   request = mock_server_receives_command (
+      server, "admin", MONGOC_QUERY_SECONDARY_OK, "{'ping': 1}");
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_QUERY);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Create a connection on client_opmsg. */
+   future = future_client_command_simple (client_opmsg,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   /* The first command on a pooled client creates a new connection. */
+   request = mock_server_receives_legacy_hello (server, NULL);
+   mock_server_replies_simple (request, HELLO_POST_OPMSG);
+   request_destroy (request);
+   /* Check that the mock server recieves an OP_MSG. */
+   request = mock_server_receives_msg (server, 0, tmp_bson ("{'ping': 1}"));
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_MSG);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Re-use client_opquery, to ensure it still uses opquery. */
+   future = future_client_command_simple (client_opquery,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   request = mock_server_receives_command (
+      server, "admin", MONGOC_QUERY_SECONDARY_OK, "{'ping': 1}");
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_QUERY);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Re-use client_opmsg, to ensure it still uses opmsg. */
+   future = future_client_command_simple (client_opmsg,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   request = mock_server_receives_msg (server, 0, tmp_bson ("{'ping': 1}"));
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_MSG);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Check that selecting the server still returns the OP_MSG server */
```

Comment:
I don't follow this comment. If you're asserting below that the maxWireVersion is 6, isn't the OP_QUERY server being returned here?

Otherwise, I'm a bit confused as to why selecting a server with `client_opquery` would return a server pertaining to `client_opmsg`.

Reply:
That was a typo. I have updated it to say `OP_QUERY`.

> Otherwise, I'm a bit confused as to why selecting a server with client_opquery would return a server pertaining to client_opmsg.

To clarify, it does not matter which client is used. `mongoc_client_select_server` should return the server description from the shared topology description. I changed this to `client_opmsg` to ensure it does not matter which client is used.